File: sdk/typescript/src/host-service.ts
import { connect } from "node:net";

import type { Interceptor, Transport } from "@connectrpc/connect";
import { createGrpcTransport, Http2SessionManager } from "@connectrpc/connect-node";

export const ENV_HOST_SERVICE_SOCKET = "GESTALT_HOST_SERVICE_SOCKET";
export const ENV_HOST_SERVICE_TOKEN = "GESTALT_HOST_SERVICE_TOKEN";
export const HOST_SERVICE_RELAY_TOKEN_HEADER =
  "x-gestalt-host-service-relay-token";
export const HOST_SERVICE_BINDING_HEADER = "x-gestalt-host-binding";
export const INVOCATION_TOKEN_HEADER = "x-gestalt-invocation-token";

export type HostServiceTransportOptions = {
  baseUrl: string;
  nodeOptions?: { path: string };
};

export type HostServiceGrpcTransport = Transport & {
  close(): void;
};

export function parseHostServiceTarget(
  serviceName: string,
  rawTarget: string,
): HostServiceTransportOptions {
  const target = rawTarget.trim();
  if (!target) {
    throw new Error(`${serviceName}: transport target is required`);
  }
  if (target.startsWith("tcp://")) {
    const address = target.slice("tcp://".length).trim();
    if (!address) {
      throw new Error(
        `${serviceName}: tcp target ${JSON.stringify(rawTarget)} is missing host:port`,
      );
    }
    return { baseUrl: `http://${address}` };
  }
  if (target.startsWith("tls://")) {
    const address = target.slice("tls://".length).trim();
    if (!address) {
      throw new Error(
        `${serviceName}: tls target ${JSON.stringify(rawTarget)} is missing host:port`,
      );
    }
    return { baseUrl: `https://${address}` };
  }
  if (target.startsWith("unix://")) {
    const socketPath = target.slice("unix://".length).trim();
    if (!socketPath) {
      throw new Error(
        `${serviceName}: unix target ${JSON.stringify(rawTarget)} is missing a socket path`,
      );
    }
    return { baseUrl: "http://localhost", nodeOptions: { path: socketPath } };
  }
  if (target.includes("://")) {
    const parsed = new URL(target);
    throw new Error(
      `${serviceName}: unsupported target scheme ${JSON.stringify(parsed.protocol.replace(/:$/, ""))}`,
    );
  }
  return { baseUrl: "http://localhost", nodeOptions: { path: target } };
}

export function hostServiceMetadataInterceptors(
  token: string,
  binding: string,
  invocationToken = "",
): Interceptor[] {
  const normalizedToken = token.trim();
  const normalizedBinding = binding.trim();
  const normalizedInvocationToken = invocationToken.trim();
  if (!normalizedToken && !normalizedBinding && !normalizedInvocationToken) {
    return [];
  }
  return [
    (next) => async (req) => {
      if (normalizedToken) {
        req.header.set(HOST_SERVICE_RELAY_TOKEN_HEADER, normalizedToken);
      }
      if (normalizedBinding) {
        req.header.set(HOST_SERVICE_BINDING_HEADER, normalizedBinding);
      }
      if (normalizedInvocationToken) {
        req.header.set(INVOCATION_TOKEN_HEADER, normalizedInvocationToken);
      }
      return await next(req);
    },
  ];
}

export function createHostServiceGrpcTransport(
  transportOptions: HostServiceTransportOptions,
  interceptors: Interceptor[] = [],
): HostServiceGrpcTransport {
  const nodeOptions = transportOptions.nodeOptions
    ? {
        createConnection: () =>
          connect({ path: transportOptions.nodeOptions!.path }),
      }
    : undefined;
  const sessionManager = new Http2SessionManager(
    transportOptions.baseUrl,
    undefined,
    nodeOptions,
  );
  const transport = createGrpcTransport({
    baseUrl: transportOptions.baseUrl,
    sessionManager,
    interceptors,
  });
  return Object.assign(transport, {
    close: () => sessionManager.abort(),
  });
}

export function requireHostServiceTarget(
  serviceName: string,
): { target: string; token: string } {
  const target = process.env[ENV_HOST_SERVICE_SOCKET];
  if (!target) {
    throw new Error(`${serviceName}: ${ENV_HOST_SERVICE_SOCKET} is not set`);
  }
  return {
    target,
    token: process.env[ENV_HOST_SERVICE_TOKEN]?.trim() ?? "",
  };
}

File: sdk/rust/src/host_service.rs
use hyper_util::rt::TokioIo;
use tokio::net::UnixStream;
use tonic::Request as GrpcRequest;
use tonic::metadata::MetadataValue;
use tonic::service::Interceptor;
use tonic::service::interceptor::InterceptedService;
use tonic::transport::{Channel, ClientTlsConfig, Endpoint, Uri};
use tower::service_fn;

use crate::env::HOST_SERVICE_BINDING_HEADER;

const RELAY_TOKEN_HEADER: &str = "x-gestalt-host-service-relay-token";

pub(crate) type Transport = InterceptedService<Channel, RelayTokenInterceptor>;

#[derive(Debug, thiserror::Error)]
pub(crate) enum HostServiceError {
    #[error("{0}")]
    Transport(#[from] tonic::transport::Error),
    #[error("{0}")]
    Env(String),
}

pub(crate) async fn connect(
    service_name: &str,
    raw_target: &str,
    relay_token: &str,
    binding: Option<&str>,
) -> Result<Transport, HostServiceError> {
    let channel = match parse_target(service_name, raw_target)? {
        Target::Unix(path) => {
            Endpoint::try_from("http://[::]:50051")?
                .connect_with_connector(service_fn(move |_: Uri| {
                    let path = path.clone();
                    async move { UnixStream::connect(path).await.map(TokioIo::new) }
                }))
                .await?
        }
        Target::Tcp(address) => {
            Endpoint::from_shared(format!("http://{address}"))?
                .connect()
                .await?
        }
        Target::Tls(address) => {
            Endpoint::from_shared(format!("https://{address}"))?
                .tls_config(ClientTlsConfig::new().with_native_roots())?
                .connect()
                .await?
        }
    };

    Ok(InterceptedService::new(
        channel,
        relay_token_interceptor(service_name, relay_token, binding)?,
    ))
}

enum Target {
    Unix(String),
    Tcp(String),
    Tls(String),
}

fn parse_target(service_name: &str, raw_target: &str) -> Result<Target, HostServiceError> {
    let target = raw_target.trim();
    if target.is_empty() {
        return Err(HostServiceError::Env(format!(
            "{service_name}: transport target is required"
        )));
    }
    if let Some(address) = target.strip_prefix("tcp://") {
        let address = address.trim();
        if address.is_empty() {
            return Err(HostServiceError::Env(format!(
                "{service_name}: tcp target {raw_target:?} is missing host:port"
            )));
        }
        return Ok(Target::Tcp(address.to_string()));
    }
    if let Some(address) = target.strip_prefix("tls://") {
        let address = address.trim();
        if address.is_empty() {
            return Err(HostServiceError::Env(format!(
                "{service_name}: tls target {raw_target:?} is missing host:port"
            )));
        }
        return Ok(Target::Tls(address.to_string()));
    }
    if let Some(path) = target.strip_prefix("unix://") {
        let path = path.trim();
        if path.is_empty() {
            return Err(HostServiceError::Env(format!(
                "{service_name}: unix target {raw_target:?} is missing a socket path"
            )));
        }
        return Ok(Target::Unix(path.to_string()));
    }
    if target.contains("://") {
        let scheme = target.split("://").next().unwrap_or_default();
        return Err(HostServiceError::Env(format!(
            "{service_name}: unsupported target scheme {scheme:?}"
        )));
    }
    Ok(Target::Unix(target.to_string()))
}

fn relay_token_interceptor(
    service_name: &str,
    token: &str,
    binding: Option<&str>,
) -> Result<RelayTokenInterceptor, HostServiceError> {
    let relay_token = if token.trim().is_empty() {
        None
    } else {
        Some(MetadataValue::try_from(token.to_string()).map_err(|err| {
            HostServiceError::Env(format!(
                "{service_name}: invalid relay token metadata: {err}"
            ))
        })?)
    };
    let binding = binding
        .filter(|value| !value.trim().is_empty())
        .map(|value| {
            MetadataValue::try_from(value.to_string()).map_err(|err| {
                HostServiceError::Env(format!("{service_name}: invalid binding metadata: {err}"))
            })
        })
        .transpose()?;

    Ok(RelayTokenInterceptor {
        relay_token,
        binding,
    })
}

#[derive(Clone)]
pub(crate) struct RelayTokenInterceptor {
    relay_token: Option<MetadataValue<tonic::metadata::Ascii>>,
    binding: Option<MetadataValue<tonic::metadata::Ascii>>,
}

impl Interceptor for RelayTokenInterceptor {
    fn call(&mut self, mut request: GrpcRequest<()>) -> Result<GrpcRequest<()>, tonic::Status> {
        if let Some(header) = self.relay_token.clone() {
            request.metadata_mut().insert(RELAY_TOKEN_HEADER, header);
        }
        if let Some(header) = self.binding.clone() {
            request
                .metadata_mut()
                .insert(HOST_SERVICE_BINDING_HEADER, header);
        }
        Ok(request)
    }
}
